File: src/main/java/game/lottery/LotteryGame.java
package game.lottery;

import java.util.Arrays;

/**
 * Lottery game.
 * 
 * @author Moshe
 */
public class LotteryGame implements Comparable<LotteryGame> {

	public int id;
	public int[] lot;
	public int bonus;

	/**
	 * Constructs a new game.
	 */
	public LotteryGame() {
		super();
	}

	/**
	 * Constructs a new game.
	 */
	public LotteryGame(int size) {
		super();
		this.lot = new int[size];
	}

	@Override
	public int hashCode() {
		return id;
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof LotteryGame) {
			LotteryGame that = (LotteryGame) obj;
			return Arrays.equals(this.lot, that.lot) && (this.bonus == that.bonus);
		}
		return super.equals(obj);
	}

	public int compareTo(LotteryGame that) {
		int size0 = this.lot.length;
		int size1 = that.lot.length;
		int comp = size0 - size1;
		if (comp == 0) {
			int[] lot0 = this.lot;
			int[] lot1 = that.lot;
			for (int i = 0; (i < size0) && (comp == 0); i++) {
				comp = lot0[i] - lot1[i];
			}

			if (comp == 0) {
				comp = this.bonus - that.bonus;
				if (comp == 0) {
					comp = this.id - that.id;
				}
			}
		}
		return comp;
	}

	@Override
	public String toString() {
		StringBuffer buf = new StringBuffer();
		buf.append(id).append(": ");
		for (int ball : lot) {
			buf.append(ball).append(' ');
		}
		buf.append('+').append(bonus);
		return buf.toString();
	}
}

File: src/main/java/com/github/pnemonic/game/lottery/LotteryGame.kt
package com.github.pnemonic.game.lottery

import com.github.game.GameOfChanceResult
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Calendar

/**
 * Lottery game results.
 */
open class LotteryGame(size: Int) : GameOfChanceResult(), Comparable<LotteryGame> {
    var date: Calendar? = null
    var id: Int = 0
    val balls: IntArray = IntArray(size)
    var bonus: Int = 0

    override fun toString(): String {
        val buf = StringBuilder()
        date?.let { buf.append(format.format(it.time)).append(' ') }
        buf.append(id).append(": ")
        for (ball in balls) {
            buf.append(ball).append(' ')
        }
        buf.append('+').append(bonus)
        return buf.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is LotteryGame) return false
        return (this.balls contentEquals other.balls) && (bonus == other.bonus)
    }

    override fun hashCode(): Int {
        return id
    }

    override fun compareTo(other: LotteryGame): Int {
        val size0 = balls.size
        val size1 = other.balls.size
        var c = size0 - size1
        if (c == 0) {
            val lot0 = balls
            val lot1 = other.balls
            var i = 0
            while (i < size0 && c == 0) {
                c = lot0[i] - lot1[i]
                i++
            }
            if (c == 0) {
                c = bonus - other.bonus
                if (c == 0) {
                    c = id - other.id
                }
            }
        }
        return c
    }

    fun copy(): LotteryGame {
        val record = this
        return LotteryGame(record.balls.size).apply {
            this.date = record.date
            this.id =  record.id
            record.balls.copyInto(this.balls)
            this.bonus = record.bonus
        }
    }

    companion object {
        private val format: DateFormat = SimpleDateFormat()
    }
}